Closes #9183, #9199: Move SessionStorage from browser-session to browser-session-storage.

pocmo opened the pull request:
Closes #9198 too. And also closes #8600. And now closes #9292 too.

This is one piece of browser-session that we want to keep and moving it out allows more components to
not depend on browser-session anymore.

Multiple things are in this patch:
* Storage related code moved from browser-session to browser-session-storage
* BrowserStateSerializer gets split into BrowserStateWriter and BrowserStateReader
* SessionManager.Snapshot has been removed and we now save and restore independened from SessionManager.
  We already saved BrowserState. Now we restore into a generic `BrowsingSession`.
* feature-tab-collections no longer depends on browser-session.
* All restore code now goes through TabsUseCases instead of SessionManager directly.

pocmo commented:
Do not land: This is for 2021. :)

pocmo commented on components/browser/session-storage/src/androidTest/java/RestoringBrowsingSessionsTest.kt:
```diff
@@ -0,0 +1,83 @@
+     * App: Sample Browser
+     * Version: 2020-12-16
+     * GeckoView: 84.0 Beta (3147bf02ca678985f814ea2b091a194a9ac22f71)
+     */
```
I want to add some more cases here and grab some serialized states from different versions of Fenix.

csadilek replied:
🔥 awesome to have those tests.

pocmo commented:
Matching Fenix changes:
https://github.com/mozilla-mobile/fenix/pull/17120

csadilek commented on .buildconfig.yml:
```diff
@@ -223,6 +223,10 @@ projects:
     publish: true
+  browser-session-storage:
+    path: components/browser/session-storage
+    description: 'Component for saving and restoring a browsing session.'
```
Oh I see, you're considering a browsing session to be a list of (reccoverable) tabs + selected tab ID. I found "session" to be confusing here.

csadilek commented on components/browser/session-storage/src/androidTest/java/FullRestoreTest.kt:
```diff
@@ -0,0 +1,131 @@
+import java.util.concurrent.TimeoutException
+
+@RunWith(AndroidJUnit4::class)
+class FullRestoreTest {
```
nit: Let's put this in a package. It's currently it root.

csadilek commented on components/browser/session-storage/src/androidTest/java/RestoringBrowsingSessionsTest.kt:
```diff
@@ -0,0 +1,83 @@
+ */
+@Suppress("TestFunctionName")
+@RunWith(AndroidJUnit4::class)
+class RestoringBrowsingSessionsTest {
```
nit: package

csadilek commented on components/browser/session-storage/src/androidTest/java/FullRestoreTest.kt:
```diff
@@ -0,0 +1,131 @@
+
+        run {
+            // -------------------------------------------------------------------------------------
+            // Restore into new classes
```
nit: new classes?

csadilek commented on components/browser/session-storage/src/main/java/mozilla/components/browser/session/storage/BrowsingSession.kt:
```diff
@@ -0,0 +1,15 @@
+/**
+ * A restored browsing session, read from disk.
+ */
+data class BrowsingSession(
```
What do you think about renaming this to `RecoverableBrowserState`, similar to `RecoverableTab`? `BrowsingSession` keeps confusing me.

pocmo replied:
Yeah, makes sense. For the other cases I wonder if we should slowly go away from "Session" and use "Tab".

csadilek commented on components/browser/session-storage/src/test/java/mozilla/components/browser/session/storage/serialize/BrowserStateWriterReaderTest.kt:
```diff
@@ -41,22 +37,24 @@ class BrowserStateSerializerTest {
         )
 
-        assertTrue(serializer.writeTab(tab, file))
+        Assert.assertTrue(writer.writeTab(tab, file))
```
nit: static import `Assert.assertTrue`, in all the asserts below too.

csadilek commented on components/feature/tabs/src/main/java/mozilla/components/feature/tabs/TabsUseCases.kt:
```diff
@@ -299,16 +303,32 @@ class TabsUseCases(
+            if (browsingSession != null) {
+                withContext(Dispatchers.Main) {
+                    invoke(browsingSession)
+                }
```
Nice! This will also close a ticket a filed a while ago: https://github.com/mozilla-mobile/android-components/issues/8600.


Can we also dispatch `RestoreCompleteAction` here which Fenix does itself right now?

csadilek commented on samples/browser/src/main/java/org/mozilla/samples/browser/SampleApplication.kt:
```diff
@@ -55,6 +56,7 @@ class SampleApplication : Application() {
         Facts.registerProcessor(LogFactProcessor())
 
         components.engine.warmUp()
+        restoreBrowsingSession()
```
nit: `restoreBrowserState` would work here too, depending how you decide above.

csadilek reviewed (commented):
Nice! 🔥 

Just made a few suggestions.

MihaiTabara reviewed (commented):
LGTM from Taskcluster perspective, looking at `taskcluster/ci/config.yml` and `.buildconfig.yml`. That should automatically inject the component in the build + sign + beetmover jobs.

I was worried about the length name of the component but I see it's 24, below the treshold from [here](https://github.com/mozilla-mobile/android-components/blob/master/taskcluster/ci/build/kind.yml#L128) so I think we should be good. CC @escapewindow

csadilek commented on components/browser/session-storage/src/main/java/mozilla/components/browser/session/storage/serialize/BrowserStateWriter.kt:
```diff
@@ -0,0 +1,118 @@
+            name(Keys.SESSION_READER_MODE_ACTIVE_URL_KEY)
+            value(tab.readerState.activeUrl)
+        }
+
```
I think we also need to re-introduce the `Session.Source` field which we seem to have dropped relative to `SessionManager`? We did add it to the store as `SessionState.Source` but we're no longer serializing/deserializing it here.

See also: https://github.com/mozilla-mobile/android-components/issues/9292

csadilek replied:
Put up a PR: https://github.com/mozilla-mobile/android-components/pull/9293

pocmo commented on components/browser/session-storage/build.gradle:
```diff
@@ -0,0 +1,77 @@
+    implementation Dependencies.androidx_lifecycle_process
+
+    // We expose this as API because we are using Observable in our public API and do not want every
+    // consumer to have to manually import "support-base".
```
Correct! I started with the `browser-session` build.gradle file for this component. Will remove this.

csadilek commented on components/browser/session-storage/src/main/java/mozilla/components/browser/session/storage/serialize/BrowserStateReader.kt:
```diff
@@ -60,76 +57,62 @@ class BrowserStateSerializer {
-    state.tabs.filter { !it.content.private }.forEachIndexed { _, tab ->
-        tab(tab)
+    while (hasNext()) {
+        when (nextName()) {
```
Hm we're having a strange crash caused by `nextName`: https://sentry.prod.mozaws.net/operations/firefox-nightly/issues/10295333/

This really shouldn't happen according to docs and after checking `hasNext` 🤔

pocmo replied:
I can't currently open that. Can you post part of the stack? In what code is it happening? (Since this here hasn't landed yet)

csadilek replied:
In our `BrowserStateSerializer`:

```
java.lang.IllegalStateException: Expected a name but was END_OBJECT
    at android.util.JsonReader.nextName(JsonReader.java:390)
    at mozilla.components.browser.session.storage.BrowserStateSerializerKt.access$tab(BrowserStateSerializer.kt:36)
    at mozilla.components.feature.tab.collections.adapter.TabAdapter.restore(TabAdapter.kt:7)
    at org.mozilla.fenix.home.sessioncontrol.DefaultSessionControlController.handleCollectionOpenTabClicked(SessionControlController.kt:8)
    at org.mozilla.fenix.home.sessioncontrol.SessionControlInteractor.onCollectionOpenTabClicked(SessionControlInteractor.kt:1)
    at org.mozilla.fenix.home.sessioncontrol.viewholders.-$$LambdaGroup$js$CmLMMZGdtRGl-qY2EQsQLG5w5hQ.onClick(java-style lambda group:1)
    at android.view.View.performClick(View.java:5646)
    at android.view.View$PerformClick.run(View.java:22473)
    at android.os.Handler.handleCallback(Handler.java:761)
    at android.os.Handler.dispatchMessage(Handler.java:98)
    at android.os.Looper.loop(Looper.java:156)
    at android.app.ActivityThread.main(ActivityThread.java:6517)
    at java.lang.reflect.Method.invoke(Method.java)
    at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:942)
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:832)
```

csadilek replied:
This happens in latest Nightly too and I can't find any usage of `nextName` that doesn't check for `hasNext`...so it's  weird.

pocmo replied:
Does it happen often? Would we want to block this on it?

csadilek replied:
Not often 85 events / 28 users. Nothing obvious in the metrics. All kinds of different Android versions and devices.

csadilek replied:
One thing I noticed that could cause this is that we don't guard against concurrent access in tab-collections. In `SessionStorage` and everywhere else we use `AtomicFile`s we guard using a lock, but we don't do that for collections? This seems like a bug? We can fix / investigate separately.

pocmo replied:
Yeah, sounds like something we could investigate. Not sure how that would actually happen when using the app, but since it is low in volume maybe some users were able to pull it off.

cadeyrn replied:
@csadilek @pocmo Please see https://github.com/mozilla-mobile/fenix/issues/17219 - except of the "nextName" line the stack trace looks very similar. And in the Firefox support we also got a similar report from another user. In both cases the crash only happens if you try to access a collection that was created before a certain build with a newer build and it's not reproducible with newly created collections.

csadilek replied:
@cadeyrn I think what users have been experiencing here most likely is https://github.com/mozilla-mobile/android-components/issues/9292 which also lines up with the stacktrace in the ticket you linked. This should already be fixed in latest Nightly.

pocmo commented:
Addressed all review comments and added additional test cases.

csadilek commented on components/browser/session-storage/src/main/java/mozilla/components/browser/session/storage/RecoverableBrowserState.kt:
```diff
@@ -0,0 +1,15 @@
+import mozilla.components.browser.state.state.recover.RecoverableTab
+
+/**
+ * A restored browsing session, read from disk.
```
nit: A restored browser state...

csadilek reviewed (approved):
👍 

One nit and a follow-up below.

pocmo commented:
Was doing one last test in Fenix and surprise.... I got this:

```Kotlin
E  Uncaught exception handled:
E  java.lang.IllegalStateException: Expected a name but was END_OBJECT
E      at android.util.JsonReader.nextName(JsonReader.java:390)
E      at mozilla.components.browser.engine.gecko.GeckoEngineSessionState$Companion.from(GeckoEngineSessionState.kt:58)
E      at mozilla.components.browser.engine.gecko.GeckoEngine.createSessionStateFrom(GeckoEngine.kt:172)
E      at mozilla.components.browser.session.storage.serialize.BrowserStateReaderKt.tab(BrowserStateReader.kt:131)
E      at mozilla.components.browser.session.storage.serialize.BrowserStateReaderKt.tabs(BrowserStateReader.kt:107)
E      at mozilla.components.browser.session.storage.serialize.BrowserStateReaderKt.browsingSession(BrowserStateReader.kt:78)
E      at mozilla.components.browser.session.storage.serialize.BrowserStateReaderKt.access$browsingSession(BrowserStateReader.kt:1)
E      at mozilla.components.browser.session.storage.serialize.BrowserStateReader.read(BrowserStateReader.kt:34)
E      at mozilla.components.browser.session.storage.SessionStorage.restore(SessionStorage.kt:48)
E      at mozilla.components.feature.tabs.TabsUseCases$RestoreUseCase$invoke$2.invokeSuspend(TabsUseCases.kt:329)
E      at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
E      at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
E      at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
E      at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
E      at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
E      at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
```

Will try to figure out how this happened.

pocmo commented:
Extracted the session storage, created a test from it and fixed it. This could have happened in situations where we couldn't serialize an engine session yet and the JSON contained: `"engineSession": {}`

pocmo commented:
Looks good now, here and testing in Fenix.
